sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala

```diff
@@ -84,11 +85,25 @@ trait FileScan extends Scan
 
   protected def seqToString(seq: Seq[Any]): String = seq.mkString("[", ", ", "]")
 
+  private lazy val (normalizedPartitionFilters, normalizedDataFilters) = {
+    val output = readSchema().toAttributes.map(a => a.withName(normalizeName(a.name)))
```

Comment:
Thinking about it again, the `FileScan` equality already considers `fileIndex` and `readSchema`, which means 2 file scans only equal to each other if they read the same set of files and the same set of columns.

Given that, I think the expr IDs do not matter for filters, only the column name matters. For normal v2 sources, they use `Filter` not `Expression`, which do not have expr IDs either.

The data/partition filters are created in `PruneFileSourcePartitions`(see https://github.com/apache/spark/blob/v3.1.1/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala#L51), and the column names inside filters are already normalized w.r.t. the actual file scan output schema, so we don't need to consider case sensitivity here.

That said, I think the normalize logic here should be very simple: just turn expr IDs to 0.

Reply:
I see your point and agree that name is that matters in these `Filter` like `Expressions` but if we go this way then I think:
- we also need to clear other properties of `AttributeReference`s like `qualifier`
- we need to either explicitly sort `partitionFilters` and `dataFilters` expression lists (probably with `.sortBy(_.hashCode())`) to make sure they match with `f.partitionFilters` and `f.dataFilters`, or use `Set(partitionFilters) == Set(f.partitionFilters)` because we can't use `ExpressionSet(partitionFilters) == ExpressionSet(f.partitionFilters)` as we removed all expr ids
- we need to reorder all descendants of each `partitionFilters` and `dataFilters` expression (with `Canonicalize.expressionReorder()` to make sure like `id = 1` matches with `1 = id` (and `Canonicalize.ignoreTimeZone()` also needs to be applied)

And just a side note that I think we could do most of the above at https://github.com/apache/spark/blob/v3.1.1/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala#L120-L121 before `withFilters()` and then `FileScan.equals()` became very simple.

But I wonder all these changes are simpler than the current PR?

Reply:
Good point. For simplicity maybe we should just assign fresh expr IDs like what this PR did, but we can remove the case sensitivity handling here.

Reply:
Ok, removed in https://github.com/apache/spark/pull/31848/commits/30d2d8b38ae95661d9461e2f6fb961e01d144e64